counsel.el

```diff
@@ -5087,6 +5088,30 @@ this if you want to provide specific non-standard build types to
 `counsel-compile'.  The default helpers are setup to handle common
 build environments.")
 
+;; Should this be hook-able to deal with project managers and other
+;; version control systems?
+(defun counsel--find-project-root ()
+  "Traverse up the tree and find the project root."
+  (expand-file-name (locate-dominating-file
+                     (or (buffer-file-name) default-directory)
+                     ".git")))
+
+;; No easy way to make directory local, would buffer local make more sense?
+(defun counsel-compile-get-filtered-history ()
+  "Return the previous compile history with non-relevant paths
+filtered out."
+  (let ((root (counsel--find-project-root))
+        (path-re (rx "cd " (group (one-or-more (not blank))) blank))
+        (kept-history))
+    (mapc
+     (lambda (hist)
+       (when (string-match path-re hist)
+         (when (string-match-p
+                root (match-string-no-properties 1 hist))
+           (push hist kept-history))))
+     counsel-compile-history)
+    kept-history))
```

Comment:
Could you elaborate why there must be a match with `cd ...`?  Will it be by convention that each compilation command will begin with `cd ...`?

Reply:
And what if there are blanks but enclosed in quotes?

Reply:
And yes, it looks hacky already.

Reply:
Yeah currently it's looking for cd foo because when we generate the make commands we prepend the cd up above. I guess a more general approach would be to regex path like structures, test if they exist and then if they do are they visible in the current project path.

Reply:
It's probably reasonable to either return a list containing for string of the command and a string of the path, which is checked separately, and used to `let`-bind `default-directory` prior invocation instead of `cd`.  Or if you don't want to return a list, then propertize the command string itself with a path (but the rest of the logic would remain the same).

Reply:
For example, in `swiper-history` I already see a similar approach is used, e.g.:

```
 #("is in state" 0 11
   (ivy-index 2))
```

That is search string stored there are propertized with some `ivy-index`.

Reply:
In any case, that would look more robust than both `cd`ing and also parsing that `cd`.

Reply:
Ahh I forgot you can use text properties to hide things. I'll look to doing that.